In [1]:
#import libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import streamlit as st
import plotly.express as px
import altair as alt

Imported the necessary libraries

In [2]:
# Load the data
df = pd.read_csv('vehicles_us.csv')

Downloaded the dataset

In [3]:
# Print general summary of the information in the dataset
df.info()
print('')
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB



price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [57]:
# Check for duplicates
duplicates = df.duplicated().sum()
print('Number of duplicates: ', duplicates)

Number of duplicates:  0


The initial exploration of the data shows the following:
There are 51525 entries in this data set

There are 13 columns in this data set that are as follows:

    Price- price of the car

    model_year - year of the cars

    model- model of the car

    condition- condition of the car

    cylinders- number of engine cylinders

    fuel - type of fuel

    odometer- mileage on the vehicle at the time of the ad

    transmission- type of transmission

    paint_color- color of the car

    is-4wd- if the car has 4 wheel drive

    date_posted- the date the ad was posted

    days_listed- the number of days the ad ran 

Entries are missing for year_model, cylinders, odometer, paint_color, and is-4wd

dtypes include: float64(4), int64(2), and object(7)

No duplicate entries were found
    

In [4]:
display(df.head(10))

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
5  14990      2014.0    chrysler 300  excellent        6.0  gas   57954.0   
6  12990      2015.0    toyota camry  excellent        4.0  gas   79212.0   
7  15990      2013.0     honda pilot  excellent        6.0  gas  109473.0   
8  11500      2012.0     kia sorento  excellent        4.0  gas  104174.0   
9   9200      2008.0     honda pilot  excellent        NaN  gas  147191.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28  
5    automatic   sedan       black     1.0  2018-06-20           15  
6    automatic   sedan       white     NaN  2018-12-27           73  
7    automatic     SUV       black     1.0  2019-01-07           68  
8    automatic     SUV         NaN     1.0  2018-07-16           19  
9    automatic     SUV        blue     1.0  2019-02-15           17

In [5]:
# Explore distribution of values in every column
for column in df.columns:
    print(df[column].value_counts(dropna=False))

price
1        798
6995     719
5995     655
4995     624
3500     620
        ... 
47665      1
22545      1
41700      1
6509       1
7670       1
Name: count, Length: 3443, dtype: int64
model_year
NaN       3619
2013.0    3549
2012.0    3468
2014.0    3448
2011.0    3375
          ... 
1948.0       1
1961.0       1
1936.0       1
1949.0       1
1929.0       1
Name: count, Length: 69, dtype: int64
model
ford f-150                           2796
chevrolet silverado 1500             2171
ram 1500                             1750
chevrolet silverado                  1271
jeep wrangler                        1119
                                     ... 
ford f-250 super duty                 241
acura tl                              236
kia sorento                           236
nissan murano                         235
mercedes-benz benze sprinter 2500      41
Name: count, Length: 100, dtype: int64
condition
excellent    24773
good         20145
like new      4742
fair          1607
new 

In [6]:
# Check for duplicates
df.duplicated().sum()

np.int64(0)

No duplicates were found

In [7]:
# Check for missing values
df.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

Further exploration of the missing values to get a missiung value count for each column.

Next steps: Handling Missing Data

Missing data types will be handled in the following way.
For the  is_4wd, this column should be a Boolean value with 1 = True and 0 = False

For the missing paint_colors, I will replace with the string 'unknown'.

For the missing values in model_year and cylinders, I will use the median values for each model of car to replace the missing values to prevent skewing the data too much.

For the missing values in odometer, I will use the median value based on the condition of the car to replace the missing values to prevent from skewing the data too much.

In [8]:
# Get the value counts for the is_4wd column
df['is_4wd'].value_counts(dropna=False)

is_4wd
NaN    25953
1.0    25572
Name: count, dtype: int64

In [9]:
# Replace missing values in is_4wd column with 0
df['is_4wd'] = df['is_4wd'].fillna(0)

# Recheck the value counts for the is_4wd column
df['is_4wd'].value_counts(dropna=False)

is_4wd
0.0    25953
1.0    25572
Name: count, dtype: int64

All missing values have been replaced using the Boolean values 0 = False and 1 = True

In [10]:
# Check for the missing values in the paint_color column
df['paint_color'].isna().sum()

np.int64(9267)

In [11]:
# Check all the values in the paint_color column
df['paint_color'].value_counts(dropna=False)

paint_color
white     10029
NaN        9267
black      7692
silver     6244
grey       5037
blue       4475
red        4421
green      1396
brown      1223
custom     1153
yellow      255
orange      231
purple      102
Name: count, dtype: int64

In [12]:
# Fill in missing values with the string 'unknown'
df['paint_color'] = df['paint_color'].fillna('unknown')

# Recheck the value counts for the paint_color column
df['paint_color'].value_counts(dropna=False)

paint_color
white      10029
unknown     9267
black       7692
silver      6244
grey        5037
blue        4475
red         4421
green       1396
brown       1223
custom      1153
yellow       255
orange       231
purple       102
Name: count, dtype: int64

All missing values have been replaced with the string 'unknown'

In [13]:
# Check for missing values in the model_year column
df['model_year'].value_counts(dropna=False)

model_year
NaN       3619
2013.0    3549
2012.0    3468
2014.0    3448
2011.0    3375
          ... 
1948.0       1
1961.0       1
1936.0       1
1949.0       1
1929.0       1
Name: count, Length: 69, dtype: int64

In [14]:
# Check median in model_year column grouped by model
df.groupby('model')['model_year'].median()

model
acura tl             2007.0
bmw x5               2010.0
buick enclave        2012.0
cadillac escalade    2009.0
chevrolet camaro     2013.0
                      ...  
toyota sienna        2008.0
toyota tacoma        2010.0
toyota tundra        2009.0
volkswagen jetta     2012.0
volkswagen passat    2013.0
Name: model_year, Length: 100, dtype: float64

In [15]:
# Replace missing values in the model_year column with the median values for each model
df['model_year'] = df.groupby('model')['model_year'].transform(lambda x: x.fillna(x.median()))

# Recheck the missing values in the model_year column
df['model_year'].value_counts(dropna=False)

model_year
2011.0    4017
2013.0    3994
2012.0    3982
2014.0    3540
2008.0    3462
          ... 
1948.0       1
1961.0       1
1936.0       1
1949.0       1
1929.0       1
Name: count, Length: 70, dtype: int64

In [16]:
# Recheck median model_year column grouped by model
df.groupby('model')['model_year'].median()

model
acura tl             2007.0
bmw x5               2010.0
buick enclave        2012.0
cadillac escalade    2009.0
chevrolet camaro     2013.0
                      ...  
toyota sienna        2008.0
toyota tacoma        2010.0
toyota tundra        2009.0
volkswagen jetta     2012.0
volkswagen passat    2013.0
Name: model_year, Length: 100, dtype: float64

Each missing model_year has been replaced with the median year that coinciding with each model

In [17]:
# Check for missing values in the cylinder column
df['cylinders'].value_counts(dropna=False)

cylinders
8.0     15844
6.0     15700
4.0     13864
NaN      5260
10.0      549
5.0       272
3.0        34
12.0        2
Name: count, dtype: int64

In [18]:
# Check median in cylinders column grouped by model
df.groupby('model')['cylinders'].median()

model
acura tl             6.0
bmw x5               6.0
buick enclave        6.0
cadillac escalade    8.0
chevrolet camaro     6.0
                    ... 
toyota sienna        6.0
toyota tacoma        6.0
toyota tundra        8.0
volkswagen jetta     4.0
volkswagen passat    4.0
Name: cylinders, Length: 100, dtype: float64

In [19]:
# Replace missing values in the cylinders column with the median values for each model
df['cylinders'] = df.groupby('model')['cylinders'].transform(lambda x: x.fillna(x.median()))

# Recheck the missing values in the cylinders column
df['cylinders'].value_counts(dropna=False)

cylinders
8.0     17712
6.0     17435
4.0     15499
10.0      549
5.0       294
3.0        34
12.0        2
Name: count, dtype: int64

In [20]:
# Recheck median in cylinders column grouped by model
df.groupby('model')['cylinders'].median()

model
acura tl             6.0
bmw x5               6.0
buick enclave        6.0
cadillac escalade    8.0
chevrolet camaro     6.0
                    ... 
toyota sienna        6.0
toyota tacoma        6.0
toyota tundra        8.0
volkswagen jetta     4.0
volkswagen passat    4.0
Name: cylinders, Length: 100, dtype: float64

Each missing cylindar value has been replaced with the median cylinder value that coinciding with each model

In [21]:
# Check values in the odometer column
df['odometer'].value_counts(dropna=False)

odometer
NaN         7892
0.0          185
140000.0     183
120000.0     179
130000.0     178
            ... 
47750.0        1
181645.0       1
105656.0       1
138608.0       1
129644.0       1
Name: count, Length: 17763, dtype: int64

In [22]:
# Check median in odometer column grouped by condition
df.groupby('condition')['odometer'].median()

condition
excellent    104230.0
fair         181613.0
good         129000.0
like new      72982.5
new            8002.0
salvage      142673.5
Name: odometer, dtype: float64

In [23]:
# Replace missing values in the odometer column with the median values for each condition
df['odometer'] = df.groupby('condition')['odometer'].transform(lambda x: x.fillna(x.median()))

# Recheck the missing values in the odometer column
df['odometer'].value_counts(dropna=False)

odometer
104230.0    3803
129000.0    3199
72982.5      728
181613.0     237
0.0          185
            ... 
111450.0       1
47750.0        1
181645.0       1
105656.0       1
138608.0       1
Name: count, Length: 17765, dtype: int64

In [24]:
# Recheck median in odometer column grouped by condition
df.groupby('condition')['odometer'].median()

condition
excellent    104230.0
fair         181613.0
good         129000.0
like new      72982.5
new            8002.0
salvage      142673.5
Name: odometer, dtype: float64

Each missing odometer value has been replaced with the median value based on the condition

In [25]:
# Recheck for missing values
df.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

All missing values have been resolved.

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    51525 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     51525 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      51525 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   51525 non-null  object 
 10  is_4wd        51525 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


Next steps: Changing data types

Several of the column data types do not coincide with the data found in the column.

I will change the model_year from float64 to int type

I will change cylinders from float64 to int type

I will change odometer from float64 to int type

I will change is_4wd from float64 to Boolean type

I will change date_posted from object to Datetime format

In [27]:
# Change model_year, cylinders, and odometer columns to integer data type
df['model_year'] = df['model_year'].astype('int')  

df['cylinders'] = df['cylinders'].astype('int')  

df['odometer'] = df['odometer'].astype('int')  


In [28]:
# Change is_4wd column to boolean data type
df['is_4wd'] = df['is_4wd'].astype(bool)

In [29]:
# Change date_posted column to datetime data type
df['date_posted'] = pd.to_datetime(df['date_posted'], format='%Y-%m-%d')

In [30]:
# Check the data types of the columns
df.dtypes

price                    int64
model_year               int64
model                   object
condition               object
cylinders                int64
fuel                    object
odometer                 int64
transmission            object
type                    object
paint_color             object
is_4wd                    bool
date_posted     datetime64[ns]
days_listed              int64
dtype: object

All data types were successfully converted

Next Steps: Adding Columns

To assist with the exploration of the data, I am going to add three new columns to the dataframe.

The first column that I am going to add is going to be a column that shows the age of the car at the time the ad was posted.

The second column that I am going to add is going to show the average milage per year for each car.

The third column that I am going to add is going to show the make for each car.

In [31]:
# Add a new column that calculates the vehicle's age when the ad was placed
df['vehicle_age'] = df['date_posted'].dt.year - df['model_year']

In [32]:
# Add a new column that calculates the vehicle's average mileage per year rounded to the nearest whole number   
df['average_mileage'] = df['odometer'] / df['vehicle_age']
df['average_mileage'] = df['average_mileage'].round(0)

In [33]:
# Display the first few rows of the modified dataset
df.head(10)


price  model_year           model  condition  cylinders fuel  odometer  \
0   9400        2011          bmw x5       good          6  gas    145000   
1  25500        2011      ford f-150       good          6  gas     88705   
2   5500        2013  hyundai sonata   like new          4  gas    110000   
3   1500        2003      ford f-150       fair          8  gas    181613   
4  14900        2017    chrysler 200  excellent          4  gas     80903   
5  14990        2014    chrysler 300  excellent          6  gas     57954   
6  12990        2015    toyota camry  excellent          4  gas     79212   
7  15990        2013     honda pilot  excellent          6  gas    109473   
8  11500        2012     kia sorento  excellent          4  gas    104174   
9   9200        2008     honda pilot  excellent          6  gas    147191   

  transmission    type paint_color  is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown    True  2018-06-23           19   
1    automatic  pickup       white    True  2018-10-19           50   
2    automatic   sedan         red   False  2019-02-07           79   
3    automatic  pickup     unknown   False  2019-03-22            9   
4    automatic   sedan       black   False  2019-04-02           28   
5    automatic   sedan       black    True  2018-06-20           15   
6    automatic   sedan       white   False  2018-12-27           73   
7    automatic     SUV       black    True  2019-01-07           68   
8    automatic     SUV     unknown    True  2018-07-16           19   
9    automatic     SUV        blue    True  2019-02-15           17   

   vehicle_age  average_mileage  
0            7          20714.0  
1            7          12672.0  
2            6          18333.0  
3           16          11351.0  
4            2          40452.0  
5            4          14488.0  
6            3          26404.0  
7            6          18246.0  
8            6          17362.0  
9           11          13381.0

In [34]:
# display information about the dataset to check for missing values in the new columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   price            51525 non-null  int64         
 1   model_year       51525 non-null  int64         
 2   model            51525 non-null  object        
 3   condition        51525 non-null  object        
 4   cylinders        51525 non-null  int64         
 5   fuel             51525 non-null  object        
 6   odometer         51525 non-null  int64         
 7   transmission     51525 non-null  object        
 8   type             51525 non-null  object        
 9   paint_color      51525 non-null  object        
 10  is_4wd           51525 non-null  bool          
 11  date_posted      51525 non-null  datetime64[ns]
 12  days_listed      51525 non-null  int64         
 13  vehicle_age      51525 non-null  int64         
 14  average_mileage  51523 non-null  float

When looking at the data, I see that there are two missing values for my average_mileage column.

In [35]:
# Display rows with missing values in the new columns
df[df['vehicle_age'].isna()]

df[df['average_mileage'].isna()]

price  model_year         model  condition  cylinders    fuel  \
17037  45400        2018      ram 3500  excellent          6  diesel   
40489  15900        2018  toyota prius  excellent          4  hybrid   

       odometer transmission       type paint_color  is_4wd date_posted  \
17037         0    automatic     pickup       black    True  2018-12-16   
40489         0    automatic  hatchback        blue   False  2018-07-30   

       days_listed  vehicle_age  average_mileage  
17037           18            0              NaN  
40489           31            0              NaN

I see that the vehicles age is listed as 0 year. I want to replace all vehicle age values that are zero with 1. 

I want to replace all missing values for mileage with 0.

In [36]:
# Replace values of 0 in the vehicle_age column with 1
df['vehicle_age'] = df['vehicle_age'].replace(0, 1)

# Replave missing values in the average_mileage column with 0
df['average_mileage'] = df['average_mileage'].fillna(0)

In [37]:
# Check for 0 values in the vehicle_age column
df['vehicle_age'].value_counts()

vehicle_age
1     4258
7     4066
6     3948
5     3845
4     3532
      ... 
71       1
59       1
82       1
69       1
89       1
Name: count, Length: 67, dtype: int64

In [38]:
# Check for missing values in the average_mileage column
df['average_mileage'].value_counts().sum()

np.int64(51525)

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   price            51525 non-null  int64         
 1   model_year       51525 non-null  int64         
 2   model            51525 non-null  object        
 3   condition        51525 non-null  object        
 4   cylinders        51525 non-null  int64         
 5   fuel             51525 non-null  object        
 6   odometer         51525 non-null  int64         
 7   transmission     51525 non-null  object        
 8   type             51525 non-null  object        
 9   paint_color      51525 non-null  object        
 10  is_4wd           51525 non-null  bool          
 11  date_posted      51525 non-null  datetime64[ns]
 12  days_listed      51525 non-null  int64         
 13  vehicle_age      51525 non-null  int64         
 14  average_mileage  51525 non-null  float

All missing values have been replaced.

In [61]:
# Check for unique values in the model column
df['model'].value_counts(dropna=False)


model
ford f-150                           2796
chevrolet silverado 1500             2171
ram 1500                             1750
chevrolet silverado                  1271
jeep wrangler                        1119
                                     ... 
ford f-250 super duty                 241
acura tl                              236
kia sorento                           236
nissan murano                         235
mercedes-benz benze sprinter 2500      41
Name: count, Length: 100, dtype: int64

We can see that there are 100 unique model types. It may be easier to aggregate data by make instead of by model. It will also let us compare popularity of models from each make.

In [46]:
# Create a new column for the make of the vehicle
df['make'] = df['model'].str.split(' ', expand=True)[0]

# Display the first few rows of the modified dataset
df.head(10)



price  model_year           model  condition  cylinders fuel  odometer  \
0   9400        2011          bmw x5       good          6  gas    145000   
1  25500        2011      ford f-150       good          6  gas     88705   
2   5500        2013  hyundai sonata   like new          4  gas    110000   
3   1500        2003      ford f-150       fair          8  gas    181613   
4  14900        2017    chrysler 200  excellent          4  gas     80903   
5  14990        2014    chrysler 300  excellent          6  gas     57954   
6  12990        2015    toyota camry  excellent          4  gas     79212   
7  15990        2013     honda pilot  excellent          6  gas    109473   
8  11500        2012     kia sorento  excellent          4  gas    104174   
9   9200        2008     honda pilot  excellent          6  gas    147191   

  transmission    type paint_color  is_4wd date_posted  days_listed  \
0    automatic     SUV     unknown    True  2018-06-23           19   
1    automatic  pickup       white    True  2018-10-19           50   
2    automatic   sedan         red   False  2019-02-07           79   
3    automatic  pickup     unknown   False  2019-03-22            9   
4    automatic   sedan       black   False  2019-04-02           28   
5    automatic   sedan       black    True  2018-06-20           15   
6    automatic   sedan       white   False  2018-12-27           73   
7    automatic     SUV       black    True  2019-01-07           68   
8    automatic     SUV     unknown    True  2018-07-16           19   
9    automatic     SUV        blue    True  2019-02-15           17   

   vehicle_age  average_mileage      make  
0            7          20714.0       bmw  
1            7          12672.0      ford  
2            6          18333.0   hyundai  
3           16          11351.0      ford  
4            2          40452.0  chrysler  
5            4          14488.0  chrysler  
6            3          26404.0    toyota  
7            6          18246.0     honda  
8            6          17362.0       kia  
9           11          13381.0     honda

In [53]:
# Check for missing values in the make column
df['make'].isna().sum()

np.int64(0)

In [54]:
# Check the values in the make column
df['make'].value_counts(dropna=False)

make
ford             12672
chevrolet        10611
toyota            5445
honda             3485
ram               3316
jeep              3281
nissan            3208
gmc               2378
subaru            1272
dodge             1255
hyundai           1173
volkswagen         869
chrysler           838
kia                585
cadillac           322
buick              271
bmw                267
acura              236
mercedes-benz       41
Name: count, dtype: int64

In [55]:
# Check the number of unique values in the make column
df['make'].nunique()

19

The make column shows 19 unique values total. There are no missing values in our new column.

Exploratory Analysis

This analysis will involve creating a variety of tables and some calculations to explore the types of cars being advertised. 

In [66]:
# Create a pivot table to calculate the average price for each type of vehicle
pivot_table = df.pivot_table(index='type', values='price', aggfunc='mean') 

# Round the values in the pivot table to 2 decimal places
pivot_table = np.round(pivot_table, 2)      

# Display the pivot table
pivot_table


price
type                 
SUV          11149.40
bus          17135.67
convertible  14575.88
coupe        14353.44
hatchback     6868.51
mini-van      8193.18
offroad      14292.29
other        10989.71
pickup       16057.41
sedan         6965.36
truck        16734.89
van          10546.94
wagon         9088.13

In [71]:
# Create a graph to visualize the average price for each type of vehicle
fig = px.bar(pivot_table, x=pivot_table.index, y='price', title='Average Price for Each Type of Vehicle')
fig.update_layout(xaxis_title='Type of Vehicle', yaxis_title='Average Price')
st.plotly_chart(fig)

2025-02-15 11:45:50.211 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-02-15 11:45:50.212 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-02-15 11:45:50.213 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.
2025-02-15 11:45:50.213 Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


DeltaGenerator()